README.md
#KNPB 0.15

This is a test piece of software. If it crashes your computer or corrupts a file
it's your own fault. I can not and will not be held responsible for such things.

The goal of this software is to create a simple VN engine using Python and the
Pygame library. It's very basic and should not be used for any serious projects.

All story data is stored in JSON files. All options for 
the game can be found in dat/config.sec in a fairly common style for config 
files. If you don't know what you're doing *do not* edit any files! 

###Controls:

* **Any Mouse Button** - Advance Story
* **Right Arrow** - Advance Story
* **Left Arrow** - Move back in story
* **M** - Toggle music on or off
* **S** - Saves while in game
* **Q** - Quit (No confirmation, just quits immediately)
* **ESC** - Return to main menu (P to go back to game)


###Config.sec parameters:

* **devmode:** True or False, skips the intro and goes right to the main menu
* **debugmode:** True or False, provides some basic info on what the game is doing.
* **gamefps:** Accepts an int, is set to 30 by default. I wouldn't set it above 60 though.
* **volume:** A number from 0.0 to 1.0. Default is 0.2. Changes in game volume.
* **fullscreen:** True or False, enables or disables fullscreen mode 

*Do not change any other options!*

###Debug mode abbreviations:

* **CS =** Current Scene
* **CC =** Current Character
* **DN =** Sees if the game has been told to go to credits. If you see True in game something is wrong.
* **DM =** Shows if Devmode is enabled or not
* **AP =** 0 if audio is not playing, 1 if it is.
* **AM =** Shows if audio is paused or not
* **AL =** Returns current volume

###Features being worked on now:

* Multiple characters on screen
* Moving characters on screen
* Changing audio tracks in game

###Things to implement in the future:

* Fade in/out
* Sound effects
* Better error handling (handling them at all would be nice)
* Options Menu
* Info screen
* Multiple saves

If you have any questions contact me: compaqxp@gmail.com, http://www.skyllama.com


CHANGES
CHANGES...

#TEST007#
*Added intro for a logo or such
*Reorginized functions
*Added shortcut to return to main menu during game (ESC)
*Enabled sound toggle on main menu
*Fixed character placement when loading from saved game

#TEST008#
*Added credits
*Added proper into/logo
*Added text wrapping
&Lost transparent story box. need to fix!

#TEST009#
*Fixed textbox transparency
*Added rounded textbox, but it's huge and ugly
*Found a slightly better audio track

#TEST010#
*Started moving things to config file (config.sec)
*Changed titles and stuff to something more general
*Added a delay in the main loop to help control cpu useage, but this is a bandaid at best.
 Uses about 30% of one of two cores on my Pentium D and 35-40mb of ram.
*Capped FPS at 20. It's a VN, it doesn't need any more at the moment. Can be changed in config.sec
*You can now use Japanese in stories if you have an apropriate font(s).

#TEST011#
*Added more config options
*Changed how intro screen runs
*Changed intro screen
*Rebranding from KNPB to more generic engine name.
*Reorganized main loop some. Should help CPU use a bit.
*Added VideoPlay and AudioStart function. Should be easy to figure out what they do.
*Removed some redundant strings.
*Changed main screen title font to something that renders better on Windows.
*Removed TextRenderMainScreen and merged it with just the regular TextRender
&Credits do not run at a stable FPS and I'm not sure why yet.
&The menu screen needs to be totaly redone in order to make buttons
 clickable. It was an afterthought at best and isen't funtional.
&Game still spikes CPU usage sometimes. I think it has to do with how
 images are being loaded. CPU spiking is worse on Linux.
 
#TEST012#
*Update readme
*Added license
*Intro can only ever be played once now, it can't be called again.
*Removed/Made comments in source more readible.
*Reorginized code so it's more readible.
*Added BGload/CharLoad funtion.
*Sound control setting added to config.sec
*After a ton of work CPU useage has been fully addressed! The game runs stable
 using under 10% of my CPU (2.13Ghz C2D P7450) 90% of the time under Windows.
 This is much better than the 30% average (with spikes to 60%) from before.
*Intro and credits run at a more stable speed now with fewer dropped frames.
*Removed time import as it's no longer needed

#TEST013#
*Added debug mode. Enable in dat/config.sec. Here is a list of debug info:
	CS = Current Scene
	CC = Current Character
	DN = Sees if the game has been told to go to credits. If you see True in game something is wrong.
	DM = Shows if Devmode is enabled or not
	AP = 0 if audio is not playing, 1 if it is.
	AM = Shows if audio is paused or not
	AL = Returns current volume
 Debug into does not show during the intro or credits. Only while playing and
 on the main menu.
 
#TEST014#
*Cleaned up variables so they all follow the same naming convention
*Cleaned up comments and fixed any spelling mistakes.
*Added option to run VN full screen or windowed. 

TODO...

*Fade in/out
*Sound effects
*Support multiple characters at once on screen
*Support changing audio track during game.
*Better error handling (handling them at all would be nice)
*Options Menu
*Info screen
*Multiple saves


vnf.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
try:
	import pygame
	import json
	import zipfile
	import io
	import pickle
	import ConfigParser
	import re
	from itertools import chain
except ImportError, message:
    raise SystemExit,  "Unable to load module. {}".format(message)
    
pygame.init()

class TextRectException:
    def __init__(self, message = None):
        self.message = message
    def __str__(self):
        return self.message
        
def TextRectRender(string, font, rect, textcolour, backgroundcolour, justification=0): #Don't touch this
    finallines = []

    requestedlines = string.splitlines()

    for requestedline in requestedlines:
        if font.size(requestedline)[0] > rect.width:
            words = requestedline.split(' ')
            for word in words:
                if font.size(word)[0] >= rect.width:
                    raise TextRectException, "The word " + word + " is too long to fit in the rect passed."
            accumulatedline = ""
            
            for word in words:
                test_line = accumulatedline + word + " "   
                if font.size(test_line)[0] < rect.width:
                    accumulatedline = test_line 
                else: 
                    finallines.append(accumulatedline) 
                    accumulatedline = word + " " 
            finallines.append(accumulatedline)
        else: 
            finallines.append(requestedline) 

    surface = pygame.Surface(rect.size, pygame.SRCALPHA) 
    surface.fill((150, 150, 150, 0))

    accumulatedheight = 0 
    for line in finallines: 
        if accumulatedheight + font.size(line)[1] >= rect.height:
            raise TextRectException, "Once wrapped, the text string was too tall to fit in the rect."
        if line != "":
            tempsurface = font.render(line, 1, textcolour)
            if justification == 0:
                surface.blit(tempsurface, (0, accumulatedheight))
            elif justification == 1:
                surface.blit(tempsurface, ((rect.width - tempsurface.get_width()) / 2, accumulatedheight))
            elif justification == 2:
                surface.blit(tempsurface, (rect.width - tempsurface.get_width(), accumulatedheight))
            else:
                raise TextRectException, "Invalid justification arg: " + str(justification)
        accumulatedheight += font.size(line)[1]

    return surface

def TextRender(text, colour, size, fontfile): #Function for rendering character and story text
	font = pygame.font.Font("fonts/{}.ttf".format(fontfile), size)
	textrender = font.render(text, 1, (colour))
	return textrender
	
def StoryData(table, value): #Parse story json file
	parsed_json = json.loads(storydata)
	json_data = parsed_json[table][value]
	return json_data
  
def LoadFromZip(location, name, filetype): #Load images, music and other content from zip file
	zipcontents = archive.read('{}/{}.{}'.format(location, name, filetype))
	bytesio = io.BytesIO(zipcontents) #Pygame won't accept files straight from zip. Do this so it can use them.
	return bytesio
	
def TransRect(width, height, r, g, b, aplha): #Used for rendering semi-transparent rects
	s = pygame.Surface((width, height), pygame.SRCALPHA)
	s.fill((r, g, b, aplha))
	return s
	
def MuteSound(audio):#Mutes game audio.
	if audio == True:
		pygame.mixer.music.pause()
		audio = False
	elif audio == False:
		pygame.mixer.music.unpause()
		audio = True
	return audio
	
def CharMove(x, y):
	characterrect.move_ip(x, y)
	return
	
def VideoPlay(videofile): #Plays videos, does nothing special.
	introvideo = pygame.movie.Movie('{}/{}.vnv'.format(avdir, videofile))
	introscreen = pygame.Surface(introvideo.get_size()).convert()
	introvideo.set_display(introscreen)
	introvideo.play()
	playing = True
	while playing:
		for event in pygame.event.get():
			if event.type == pygame.QUIT:
				introvideo.stop()
				playing = False
				done = True
				
		screen.blit(introscreen,(0,0))
		pygame.display.update()
		click.tick(int(config.get('settings', 'videofps')))
		if introvideo.get_busy() == False:
			playing = False
	return

def StartAudio(audiofile): #Starts music playing, can be called at any point to change audio.
	pygame.mixer.init()
	pygame.mixer.music.load("{}/{}.vnm".format(avdir, audiofile))
	pygame.mixer.music.set_volume(float(config.get('settings', 'volume')))
	return
	
def BGLoad(): #Loads background image
	background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
	backgroundrect = background.get_rect()
	return background, backgroundrect
	
def CharLoad(): #Loads character image
	characterimage = pygame.image.load(LoadFromZip(characterdir, StoryData(table, "characterimage"), "png"))
	characterrect = characterimage.get_rect()
	CharMove(0, 0)
	return characterimage, characterrect
	
def MainMenu(): #Display main menu title/options. This is only temporary and will be redone when the rest of the game works fine.
	if mainmenurendered == False:
		screen.blit(BGLoad()[0], BGLoad()[1])
	
		MainMenuTitle = config.get('vn_info', 'mainmenutitle')
		Dev_Info = config.get('vn_info', 'devinfo')
		Copyright = config.get('vn_info', 'copyright')
		#Game title
		screen.blit(TextRender(MainMenuTitle, BLACK, 39, 'titles'), (3, 5))
		screen.blit(TextRender(Dev_Info, BLACK, 16, 'titles'), (350, 50))
		screen.blit(TextRender(Copyright, BLACK, 10, 'titles'), (600, 588))
		
		#Background for options
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (15,535)) #w h r g b a
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (172,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (329,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (486,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (643,535))
		
		#Options
		screen.blit(TextRender("(P)lay", BLACK, menubuttons, 'main'), (55, 547))
		screen.blit(TextRender("(L)oad", BLACK, menubuttons, 'main'), (210, 547))
		screen.blit(TextRender("(O)ptions", BLACK, menubuttons, 'main'), (353, 547))
		screen.blit(TextRender("(I)nfo", BLACK, menubuttons, 'main'), (527, 547))
		screen.blit(TextRender("(Q)uit", BLACK, menubuttons, 'main'), (682, 547))
	else:
		pass
	return
	
def DebugInfo(): #Puts Debug info on screen if DebugMode is set true in config.sec
	if config.get('settings', 'debug') == 'True':
		charname = re.sub('[<>]', '', character)
		screen.blit(TextRender('CS: {}'.format(currentscene), RED, 16, 'debug'), (10, 10))
		screen.blit(TextRender('CC: {}'.format(charname), RED, 16, 'debug'), (10, 25))
		screen.blit(TextRender('DN: {}'.format(done), RED, 16, 'debug'), (10, 40))
		screen.blit(TextRender('DM: {}'.format(devmode), RED, 16, 'debug'), (10, 55))
		screen.blit(TextRender('AP: {}'.format(pygame.mixer.music.get_busy()), RED, 16, 'debug'), (10, 70))
		screen.blit(TextRender('AM: {}'.format(audio), RED, 16, 'debug'), (10, 85))
		screen.blit(TextRender('AL: {}'.format(config.get('settings', 'volume')), RED, 16, 'debug'), (10, 100))
	else:
		pass
	return
	
config = ConfigParser.ConfigParser()
config.read('dat/config.sec')	

#Directory info
backgrounddir = config.get('dirs', 'backgrounddir')
characterdir = config.get('dirs', 'characterdir')
avdir = config.get('dirs', 'avdir')
uidir = config.get('dirs', 'uidir')
otherdir = config.get('dirs', 'otherdir')

#Set the zip file that contains resources
archive = zipfile.ZipFile("{}".format(config.get('files', 'assets')), "r")

#Set window size, give title and set if the game should run full screen or windowed.
width = int(config.get('settings', 'windowwidth'))
height = int(config.get('settings', 'windowheight'))
size = (width, height)
pygame.display.set_caption("{}".format(config.get('vn_info', 'windowtitle')))

if config.get('settings', 'fullscreen') == 'True':
	screen = pygame.display.set_mode(size,pygame.FULLSCREEN)
else:
	screen = pygame.display.set_mode(size)

#Commonly used colours
WHITE = (255,255,255)
BLACK = (0,0,0)
RED = (235,16,16)

#Common font sizes
charactername = int(config.get('fonts', 'charactername'))
storysize = int(config.get('fonts', 'storysize'))
menubuttons = int(config.get('fonts', 'menubuttons'))

#Set fonts
regularfont = pygame.font.Font("{}".format(config.get('files', 'regularfont')), storysize)
debugfont = config.get('files', 'debugfont')

#Open and read story contents from json file
storyconfigread = open("{}".format(config.get('files', 'storycontent')), 'r')
storydata = storyconfigread.read()

#Inital json data to grab. We need this to start the story.
table = config.get('json_info', 'initaltable')
character = StoryData(table, "character")
storytext = StoryData(table, "dialog") 

#Load first bg/char to the screen and load inital music
background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
backgroundrect = background.get_rect()
characterimage = pygame.image.load(LoadFromZip(characterdir, StoryData(table, "characterimage"), "png"))
characterrect = characterimage.get_rect()
currentscene = "A" #This is just garbage, this variable gets changed later.
click = pygame.time.Clock()
CharMove(100,150)
StartAudio('music')

#Set inital variables for the main loop
done = False
audio = True
fromgame = False
mainmenurendered = False
devmode = config.get('settings', 'devmode')

if devmode == 'False':
	menu = False
	intro = True
	
elif devmode == 'True':
	menu = True
	intro = False
	pygame.mixer.music.play()


main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
try:
	from vnf import *
except ImportError, message:
    raise SystemExit,  "Unable to load module. {}".format(message)

#Loop for the intro, this only ever needs to play once so it's not in the main loop
while intro == True:
	VideoPlay('intro')
	pygame.mixer.music.play()
	intro = False
	menu = True

#Main loop
while done == False:
	key = pygame.key.get_pressed()
	
	if menu == True: #This is the code for the main menu. It should always come up after the intro
		MainMenu()
		mainmenurendered = True
		DebugInfo()
		for event in pygame.event.get():
			if event.type == pygame.QUIT:
				done = True
				
			elif key[pygame.K_q]:
				done = True
				
			elif key[pygame.K_p]:
				menu = False
				if fromgame == True:
					background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
					backgroundrect = background.get_rect()
				
			elif key[pygame.K_l]: #load save data from previous session
				with open('dat/savefile') as f:
					data = pickle.load(f)
				table = data
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				menu = False
				
			elif key[pygame.K_m]: #Audio toggle. User can press M key to pause or unpause	
				MuteSound(audio)
				audio = MuteSound(audio)
	
	elif menu == False: #This is the code for the main game, when menu == false we can begin
		if table != currentscene: #Put char, bg and text on screen
			textbox = pygame.image.load(LoadFromZip(uidir, "textbox", "png"))
			textboxrect = textbox.get_rect()
			screen.blit(BGLoad()[0], BGLoad()[1])
			screen.blit(CharLoad()[0], CharLoad()[1])
			screen.blit(textbox, textboxrect)
			screen.blit(TextRender("{}".format(character), BLACK, charactername, 'main'), (40, 485))
		
			charactertextbox = pygame.Rect((15, 515, 780, 75))
			rendered_text = TextRectRender(storytext, regularfont, charactertextbox, (0, 0, 0), (191, 191, 191, 150), 0)
			screen.blit(rendered_text, charactertextbox.topleft)
        
			data = table
			currentscene = table
			DebugInfo()
		else:
			pass

		if pygame.mixer.music.get_busy() == False: #If music has finished playing start it again
			pygame.mixer.music.play()
	
		if StoryData(table, "end") == "True": #If the story is done we return to main menu.
			table = "scene1"
			storytext = StoryData(table, "dialog")
			character = StoryData(table, "character")
			VideoPlay('credits')
			intro = False
			playing = False
			done = False
			menu = True

		for event in pygame.event.get(): #Get any event (keybo/mouse presses)
			if event.type == pygame.QUIT:
				done = True

			#If the mouse is clicked get everything set up to advance to next scene
			elif (event.type == pygame.MOUSEBUTTONUP) or (key[pygame.K_RIGHT]):
				currentscene = table
				table = StoryData(table, "nextscene")
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				
			elif key[pygame.K_LEFT]: #If user presses left arrow fo back to previous scene
				currentscene = table
				table = StoryData(table, "lastscene")
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				
			elif key[pygame.K_m]: #Audio toggle. User can press M key to pause or unpause	
				MuteSound(audio)
				audio = MuteSound(audio)
					
			elif key[pygame.K_s]:#key for saving game
				with open('dat/savefile', 'w') as f:
					pickle.dump(data, f)
				
			elif key[pygame.K_q]: #Shortcut for quiting game. Exits right away, should ask for confirmation.
				done = True
			
			elif key[pygame.K_ESCAPE]:
				menu = True
				fromgame = True
				background = pygame.image.load(LoadFromZip(backgrounddir, "school", "jpg"))
				backgroundrect = background.get_rect()
			
	pygame.display.flip() 
	click.tick(int(config.get('settings', 'gamefps')))

pygame.quit()


Koko No Puroguramingu Basho.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
try:
	import pygame
	import json
	import zipfile
	import io
	import pickle
	import ConfigParser
	import re
	from itertools import chain
except ImportError, message:
    raise SystemExit,  "Unable to load module. {}".format(message)

pygame.init()

class TextRectException:
    def __init__(self, message = None):
        self.message = message
    def __str__(self):
        return self.message
        
def TextRectRender(string, font, rect, textcolour, backgroundcolour, justification=0): #Don't touch this
    finallines = []

    requestedlines = string.splitlines()

    for requestedline in requestedlines:
        if font.size(requestedline)[0] > rect.width:
            words = requestedline.split(' ')
            for word in words:
                if font.size(word)[0] >= rect.width:
                    raise TextRectException, "The word " + word + " is too long to fit in the rect passed."
            accumulatedline = ""
            
            for word in words:
                test_line = accumulatedline + word + " "   
                if font.size(test_line)[0] < rect.width:
                    accumulatedline = test_line 
                else: 
                    finallines.append(accumulatedline) 
                    accumulatedline = word + " " 
            finallines.append(accumulatedline)
        else: 
            finallines.append(requestedline) 

    surface = pygame.Surface(rect.size, pygame.SRCALPHA) 
    surface.fill((150, 150, 150, 0))

    accumulatedheight = 0 
    for line in finallines: 
        if accumulatedheight + font.size(line)[1] >= rect.height:
            raise TextRectException, "Once wrapped, the text string was too tall to fit in the rect."
        if line != "":
            tempsurface = font.render(line, 1, textcolour)
            if justification == 0:
                surface.blit(tempsurface, (0, accumulatedheight))
            elif justification == 1:
                surface.blit(tempsurface, ((rect.width - tempsurface.get_width()) / 2, accumulatedheight))
            elif justification == 2:
                surface.blit(tempsurface, (rect.width - tempsurface.get_width(), accumulatedheight))
            else:
                raise TextRectException, "Invalid justification arg: " + str(justification)
        accumulatedheight += font.size(line)[1]

    return surface

def TextRender(text, colour, size, fontfile): #Function for rendering character and story text
	font = pygame.font.Font("fonts/{}.ttf".format(fontfile), size)
	textrender = font.render(text, 1, (colour))
	return textrender
	
def StoryData(table, value): #Parse story json file
	parsed_json = json.loads(storydata)
	json_data = parsed_json[table][value]
	return json_data
  
def LoadFromZip(location, name, filetype): #Load images, music and other content from zip file
	zipcontents = archive.read('{}/{}.{}'.format(location, name, filetype))
	bytesio = io.BytesIO(zipcontents) #Pygame won't accept files straight from zip. Do this so it can use them.
	return bytesio
	
def TransRect(width, height, r, g, b, aplha): #Used for rendering semi-transparent rects
	s = pygame.Surface((width, height), pygame.SRCALPHA)
	s.fill((r, g, b, aplha))
	return s
	
def MuteSound(audio):#Mutes game audio.
	if audio == True:
		pygame.mixer.music.pause()
		audio = False
	elif audio == False:
		pygame.mixer.music.unpause()
		audio = True
	return audio
	
def CharMove(x, y):
	characterrect.move_ip(x, y)
	return
	
def VideoPlay(videofile): #Plays videos, does nothing special.
	introvideo = pygame.movie.Movie('{}/{}.vnv'.format(avdir, videofile))
	introscreen = pygame.Surface(introvideo.get_size()).convert()
	introvideo.set_display(introscreen)
	introvideo.play()
	playing = True
	while playing:
		for event in pygame.event.get():
			if event.type == pygame.QUIT:
				introvideo.stop()
				playing = False
				done = True
				
		screen.blit(introscreen,(0,0))
		pygame.display.update()
		click.tick(int(config.get('settings', 'videofps')))
		if introvideo.get_busy() == False:
			playing = False
	return

def StartAudio(audiofile): #Starts music playing, can be called at any point to change audio.
	pygame.mixer.init()
	pygame.mixer.music.load("{}/{}.vnm".format(avdir, audiofile))
	pygame.mixer.music.set_volume(float(config.get('settings', 'volume')))
	return
	
def BGLoad(): #Loads background image
	background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
	backgroundrect = background.get_rect()
	return background, backgroundrect
	
def CharLoad(): #Loads character image
	characterimage = pygame.image.load(LoadFromZip(characterdir, StoryData(table, "characterimage"), "png"))
	characterrect = characterimage.get_rect()
	CharMove(0, 0)
	return characterimage, characterrect
	
def MainMenu(): #Display main menu title/options. This is only temporary and will be redone when the rest of the game works fine.
	if mainmenurendered == False:
		screen.blit(BGLoad()[0], BGLoad()[1])
	
		MainMenuTitle = config.get('vn_info', 'mainmenutitle')
		Dev_Info = config.get('vn_info', 'devinfo')
		Copyright = config.get('vn_info', 'copyright')
		#Game title
		screen.blit(TextRender(MainMenuTitle, BLACK, 39, 'titles'), (3, 5))
		screen.blit(TextRender(Dev_Info, BLACK, 16, 'titles'), (350, 50))
		screen.blit(TextRender(Copyright, BLACK, 10, 'titles'), (600, 588))
		
		#Background for options
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (15,535)) #w h r g b a
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (172,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (329,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (486,535))
		screen.blit(TransRect(142, 50, 191, 191, 191, 150), (643,535))
		
		#Options
		screen.blit(TextRender("(P)lay", BLACK, menubuttons, 'main'), (55, 547))
		screen.blit(TextRender("(L)oad", BLACK, menubuttons, 'main'), (210, 547))
		screen.blit(TextRender("(O)ptions", BLACK, menubuttons, 'main'), (353, 547))
		screen.blit(TextRender("(I)nfo", BLACK, menubuttons, 'main'), (527, 547))
		screen.blit(TextRender("(Q)uit", BLACK, menubuttons, 'main'), (682, 547))
	else:
		pass
	return
	
def DebugInfo(): #Puts Debug info on screen if DebugMode is set true in config.sec
	if config.get('settings', 'debug') == 'True':
		charname = re.sub('[<>]', '', character)
		screen.blit(TextRender('CS: {}'.format(currentscene), RED, 16, 'debug'), (10, 10))
		screen.blit(TextRender('CC: {}'.format(charname), RED, 16, 'debug'), (10, 25))
		screen.blit(TextRender('DN: {}'.format(done), RED, 16, 'debug'), (10, 40))
		screen.blit(TextRender('DM: {}'.format(devmode), RED, 16, 'debug'), (10, 55))
		screen.blit(TextRender('AP: {}'.format(pygame.mixer.music.get_busy()), RED, 16, 'debug'), (10, 70))
		screen.blit(TextRender('AM: {}'.format(audio), RED, 16, 'debug'), (10, 85))
		screen.blit(TextRender('AL: {}'.format(config.get('settings', 'volume')), RED, 16, 'debug'), (10, 100))
	else:
		pass
	return
	
config = ConfigParser.ConfigParser()
config.read('dat/config.sec')	

#Directory info
backgrounddir = config.get('dirs', 'backgrounddir')
characterdir = config.get('dirs', 'characterdir')
avdir = config.get('dirs', 'avdir')
uidir = config.get('dirs', 'uidir')
otherdir = config.get('dirs', 'otherdir')

#Set the zip file that contains resources
archive = zipfile.ZipFile("{}".format(config.get('files', 'assets')), "r")

#Set window size, give title and set if the game should run full screen or windowed.
width = int(config.get('settings', 'windowwidth'))
height = int(config.get('settings', 'windowheight'))
size = (width, height)
pygame.display.set_caption("{}".format(config.get('vn_info', 'windowtitle')))

if config.get('settings', 'fullscreen') == 'True':
	screen = pygame.display.set_mode(size,pygame.FULLSCREEN)
else:
	screen = pygame.display.set_mode(size)

#Commonly used colours
WHITE = (255,255,255)
BLACK = (0,0,0)
RED = (235,16,16)

#Common font sizes
charactername = int(config.get('fonts', 'charactername'))
storysize = int(config.get('fonts', 'storysize'))
menubuttons = int(config.get('fonts', 'menubuttons'))

#Set fonts
regularfont = pygame.font.Font("{}".format(config.get('files', 'regularfont')), storysize)
debugfont = config.get('files', 'debugfont')

#Open and read story contents from json file
storyconfigread = open("{}".format(config.get('files', 'storycontent')), 'r')
storydata = storyconfigread.read()

#Inital json data to grab. We need this to start the story.
table = config.get('json_info', 'initaltable')
character = StoryData(table, "character")
storytext = StoryData(table, "dialog") 

#Load first bg/char to the screen and load inital music
background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
backgroundrect = background.get_rect()
characterimage = pygame.image.load(LoadFromZip(characterdir, StoryData(table, "characterimage"), "png"))
characterrect = characterimage.get_rect()
currentscene = "A" #This is just garbage, this variable gets changed later.
click = pygame.time.Clock()
CharMove(100,150)
StartAudio('music')

#Set inital variables for the main loop
done = False
audio = True
fromgame = False
mainmenurendered = False
devmode = config.get('settings', 'devmode')

if devmode == 'False':
	menu = False
	intro = True
	
elif devmode == 'True':
	menu = True
	intro = False
	pygame.mixer.music.play()
	
#Loop for the intro, this only ever needs to play once so it's not in the main loop
while intro == True:
	VideoPlay('intro')
	pygame.mixer.music.play()
	intro = False
	menu = True

#Main loop
while done == False:
	key = pygame.key.get_pressed()
	
	if menu == True: #This is the code for the main menu. It should always come up after the intro
		MainMenu()
		mainmenurendered = True
		DebugInfo()
		for event in pygame.event.get():
			if event.type == pygame.QUIT:
				done = True
				
			elif key[pygame.K_q]:
				done = True
				
			elif key[pygame.K_p]:
				menu = False
				if fromgame == True:
					background = pygame.image.load(LoadFromZip(backgrounddir, StoryData(table, "background"), "jpg"))
					backgroundrect = background.get_rect()
				
			elif key[pygame.K_l]: #load save data from previous session
				with open('dat/savefile') as f:
					data = pickle.load(f)
				table = data
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				menu = False
				
			elif key[pygame.K_m]: #Audio toggle. User can press M key to pause or unpause	
				MuteSound(audio)
				audio = MuteSound(audio)
	
	elif menu == False: #This is the code for the main game, when menu == false we can begin
		if table != currentscene: #Put char, bg and text on screen
			textbox = pygame.image.load(LoadFromZip(uidir, "textbox", "png"))
			textboxrect = textbox.get_rect()
			screen.blit(BGLoad()[0], BGLoad()[1])
			screen.blit(CharLoad()[0], CharLoad()[1])
			screen.blit(textbox, textboxrect)
			screen.blit(TextRender("{}".format(character), BLACK, charactername, 'main'), (40, 485))
		
			charactertextbox = pygame.Rect((15, 515, 780, 75))
			rendered_text = TextRectRender(storytext, regularfont, charactertextbox, (0, 0, 0), (191, 191, 191, 150), 0)
			screen.blit(rendered_text, charactertextbox.topleft)
        
			data = table
			currentscene = table
			DebugInfo()
		else:
			pass

		if pygame.mixer.music.get_busy() == False: #If music has finished playing start it again
			pygame.mixer.music.play()
	
		if StoryData(table, "end") == "True": #If the story is done we return to main menu.
			table = "scene1"
			storytext = StoryData(table, "dialog")
			character = StoryData(table, "character")
			VideoPlay('credits')
			intro = False
			playing = False
			done = False
			menu = True

		for event in pygame.event.get(): #Get any event (keybo/mouse presses)
			if event.type == pygame.QUIT:
				done = True

			#If the mouse is clicked get everything set up to advance to next scene
			elif (event.type == pygame.MOUSEBUTTONUP) or (key[pygame.K_RIGHT]):
				currentscene = table
				table = StoryData(table, "nextscene")
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				
			elif key[pygame.K_LEFT]: #If user presses left arrow fo back to previous scene
				currentscene = table
				table = StoryData(table, "lastscene")
				storytext = StoryData(table, "dialog")
				character = StoryData(table, "character")
				
			elif key[pygame.K_m]: #Audio toggle. User can press M key to pause or unpause	
				MuteSound(audio)
				audio = MuteSound(audio)
					
			elif key[pygame.K_s]:#key for saving game
				with open('dat/savefile', 'w') as f:
					pickle.dump(data, f)
				
			elif key[pygame.K_q]: #Shortcut for quiting game. Exits right away, should ask for confirmation.
				done = True
			
			elif key[pygame.K_ESCAPE]:
				menu = True
				fromgame = True
				background = pygame.image.load(LoadFromZip(backgrounddir, "school", "jpg"))
				backgroundrect = background.get_rect()
			
	pygame.display.flip() 
	click.tick(int(config.get('settings', 'gamefps')))

pygame.quit()
